File: typescript/packages/core/src/ops/generic/factory.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { Accessor } from '../../accessor/base.ts'
import type { OpKwargs, RegisteredOp } from '../registry.ts'
import { extractWriteData } from '../write_args.ts'
import { sliceWindow } from '../../utils/ranges.ts'
import type { PathSpec } from '../../types.ts'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type OpCoreFn = (...args: any[]) => unknown

/**
 * Structural subset of a backend's `CommandIO` the ops factory consumes.
 * The table in `commands/builtin/<b>/ops.ts` already carries every core
 * function the VFS/FUSE op wrappers forward to, so the same table feeds
 * both `makeGenericCommands` and `makeGenericOps`. Command-only fields
 * (`readStream`, `isMounted`, `find`, ...) are ignored.
 */
export interface OpsTable<A extends Accessor = Accessor> {
  readdir: (accessor: A, path: PathSpec, index?: OpKwargs['index']) => unknown
  readBytes: (accessor: A, path: PathSpec, index?: OpKwargs['index']) => Promise<Uint8Array>
  readRange?: (
    accessor: A,
    path: PathSpec,
    index: OpKwargs['index'],
    offset: number,
    size: number | null,
  ) => Promise<Uint8Array>
  stat: (accessor: A, path: PathSpec, index?: OpKwargs['index']) => unknown
  write?: OpCoreFn
  mkdir?: OpCoreFn
  unlink?: OpCoreFn
  rmdir?: OpCoreFn
  rename?: OpCoreFn
  create?: OpCoreFn
  truncate?: OpCoreFn
  append?: OpCoreFn
  setAttrs?: OpCoreFn
}

export interface MakeGenericOpsOptions {
  /** Synthesize truncate from readBytes + write (no native partial write). */
  emulateTruncate?: boolean
  /** Forward `parents=true` to the core mkdir (disk). */
  mkdirParents?: boolean
  /** Op names to skip because the backend registers an irregular wrapper. */
  overrides?: ReadonlySet<string>
  /**
   * Forward `kwargs.index` into read/readdir/stat (default true). Off for
   * the mutable local backends (ram/disk/redis/ssh), whose non-zero index
   * TTL would otherwise let a cached listing outlive the mutation.
   *
   * `ws.dispatch` is safe on its own: the dispatcher calls
   * `invalidateAfterWriteByPath` after every write op. `WorkspaceFS` is
   * not. It reaches `OpsRegistry.call` directly, so a forwarded index is
   * never evicted and readdir replays the pre-mutation listing for the
   * whole TTL. Python has no such knob because its `Ops` facade takes an
   * `on_write` hook (`Ops(..., on_write=self._invalidate_after_write_by_path)`)
   * and fires it on the same path; giving `WorkspaceFS` that hook is what
   * would retire this option. `packages/node/src/ops/index_invalidation.test.ts`
   * pins both halves.
   */
  forwardIndex?: boolean
}

const expectPathSpec = (value: unknown, op: string): PathSpec => {
  if (value === null || typeof value !== 'object' || !('virtual' in value)) {
    throw new TypeError(`${op} op requires a dst PathSpec as the first arg`)
  }
  return value as PathSpec
}

const expectLength = (value: unknown): number => {
  if (typeof value !== 'number') {
    throw new TypeError('truncate op requires a number length as the first arg')
  }
  return value
}

/**
 * Generate a backend's VFS/FUSE op set from its `CommandIO` table.
 *
 * The per-backend `ops/<b>/` wrapper modules were hand-written forwards
 * of a handful of shapes; this factory emits the same wrappers from the
 * table that already feeds `makeGenericCommands`, so a backend declares
 * its core surface once. Ops whose table field is undefined are
 * omitted, mirroring how the command factory skips write commands on
 * read-only backends.
 */
export function makeGenericOps<A extends Accessor>(
  resource: string | readonly string[],
  table: OpsTable<A>,
  options: MakeGenericOpsOptions = {},
): RegisteredOp[] {
  const resources = typeof resource === 'string' ? [resource] : resource
  const skip = options.overrides ?? new Set<string>()
  const ops: RegisteredOp[] = []

  const emit = (
    name: string,
    fn: RegisteredOp['fn'],
    write: boolean,
    filetype: string | null = null,
  ): void => {
    if (skip.has(name)) return
    for (const res of resources) {
      ops.push({ name, resource: res, filetype, fn, write })
    }
  }

  const asA = (accessor: Accessor): A => accessor as A
  const pickIndex = (kwargs: OpKwargs): OpKwargs['index'] =>
    options.forwardIndex === false ? undefined : kwargs.index

  // A backend that can fetch a range natively does so, which is the whole
  // point on an object store: one ranged GET instead of the whole file. Every
  // other backend falls back to reading and slicing, which is the same answer
  // at the same cost as before, and is the only meaningful behavior for a
  // backend that renders its content rather than storing it, since there is no
  // remote range to ask for. A zero-length read is answered here rather than
  // sent anywhere: no store can express an empty range, and the answer is known.
  emit(
    'read',
    async (accessor, path, _args, kwargs) => {
      const offset = typeof kwargs.offset === 'number' ? kwargs.offset : 0
      const size = typeof kwargs.size === 'number' ? kwargs.size : null
      if (size === 0) return new Uint8Array(0)
      const whole = offset === 0 && size === null
      const native = table.readRange
      if (native !== undefined && !whole) {
        return native(asA(accessor), path, pickIndex(kwargs), offset, size)
      }
      const data = await table.readBytes(asA(accessor), path, pickIndex(kwargs))
      return whole ? data : sliceWindow(data, offset, size)
    },
    false,
  )
  emit(
    'readdir',
    (accessor, path, _args, kwargs) => table.readdir(asA(accessor), path, pickIndex(kwargs)),
    false,
  )
  emit(
    'stat',
    (accessor, path, _args, kwargs) => table.stat(asA(accessor), path, pickIndex(kwargs)),
    false,
  )

  const { write, mkdir, unlink, rmdir, rename, create, truncate, append, setAttrs } = table
  if (write) {
    emit(
      'write',
      (accessor, path, args) => write(asA(accessor), path, extractWriteData(args)),
      true,
    )
  }
  if (append) {
    emit(
      'append',
      (accessor, path, args) => append(asA(accessor), path, extractWriteData(args)),
      true,
    )
  }
  if (create) {
    emit('create', (accessor, path) => create(asA(accessor), path), true)
  }
  if (mkdir) {
    emit(
      'mkdir',
      (accessor, path) =>
        options.mkdirParents ? mkdir(asA(accessor), path, true) : mkdir(asA(accessor), path),
      true,
    )
  }
  if (unlink) {
    emit('unlink', (accessor, path) => unlink(asA(accessor), path), true)
  }
  if (rmdir) {
    emit('rmdir', (accessor, path) => rmdir(asA(accessor), path), true)
  }
  if (rename) {
    emit(
      'rename',
      (accessor, path, args) => rename(asA(accessor), path, expectPathSpec(args[0], 'rename')),
      true,
    )
  }

  if (truncate) {
    emit(
      'truncate',
      (accessor, path, args) => truncate(asA(accessor), path, expectLength(args[0])),
      true,
    )
  } else if (options.emulateTruncate) {
    if (!write) {
      throw new Error('emulateTruncate requires a write op on the table')
    }
    emit(
      'truncate',
      async (accessor, path, args) => {
        const length = expectLength(args[0])
        let data: Uint8Array
        try {
          data = await table.readBytes(asA(accessor), path)
        } catch (err) {
          if ((err as { code?: string }).code !== 'ENOENT') throw err
          data = new Uint8Array(0)
        }
        const out = new Uint8Array(length)
        out.set(data.subarray(0, length))
        return write(asA(accessor), path, out)
      },
      true,
    )
  }

  if (setAttrs) {
    emit('setattr', (accessor, path, _args, kwargs) => setAttrs(asA(accessor), path, kwargs), true)
  }

  return ops
}

File: python/mirage/ops/generic/factory.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.base import Accessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.ops.generic.table import OpFn, OpsTable
from mirage.ops.registry import RegisteredOp
from mirage.types import PathSpec
from mirage.utils.ranges import is_unsatisfiable_range, slice_window


def _make_read(fn: OpFn) -> OpFn:

    async def read(accessor: Accessor,
                   path: PathSpec,
                   *,
                   index: IndexCacheStore | None = None,
                   **kwargs) -> bytes:
        return await fn(accessor, path, index)

    return read


def _make_ranged_read(table: OpsTable) -> OpFn:
    """Build the ``read`` op, honoring a byte range when one is asked for.

    A backend that can fetch a range natively does so, which is the whole
    point on an object store: one ranged GET instead of the whole file.
    Every other backend falls back to reading and slicing, which is the
    same answer at the same cost as before, and is the only meaningful
    behavior for a backend that renders its content rather than storing
    it, since there is no remote range to ask for.

    A zero-length read is answered here rather than sent anywhere: no
    store can express an empty range, and the answer is known.

    A window starting at or past EOF is the one case where the two paths
    do not agree on their own: slicing yields empty, the POSIX answer,
    while an HTTP store refuses with 416. Normalizing here rather than in
    each reader keeps the op's contract one thing, and keeps a backend
    from becoming the odd one out the day it grows a native range.

    Args:
        table (OpsTable): the backend's op table.
    """

    async def read(accessor: Accessor,
                   path: PathSpec,
                   *,
                   index: IndexCacheStore | None = None,
                   offset: int = 0,
                   size: int | None = None,
                   **kwargs) -> bytes:
        if size == 0:
            return b""
        whole = not offset and size is None
        native = table.read_range
        if native is not None and not whole:
            try:
                return await native(accessor, path, index, offset, size)
            except Exception as exc:
                if not is_unsatisfiable_range(exc):
                    raise
                return b""
        data = await table.read_bytes(accessor, path, index)
        if whole:
            return data
        return slice_window(data, offset, size)

    return read


def _make_data_write(fn: OpFn) -> OpFn:

    async def write(accessor: Accessor, path: PathSpec, data: bytes,
                    **kwargs) -> None:
        await fn(accessor, path, data)

    return write


def _make_emulated_append(read_bytes: OpFn, write_bytes: OpFn) -> OpFn:

    async def append(accessor: Accessor,
                     path: PathSpec,
                     data: bytes,
                     *,
                     index: IndexCacheStore | None = None,
                     **kwargs) -> None:
        # The read takes the caller's index, like every other read here:
        # an id-addressed backend (Box, Drive) turns a path into an id
        # through it, and without one every read is a miss, so each append
        # would overwrite what the last one wrote.
        try:
            existing = await read_bytes(accessor, path, index)
        except FileNotFoundError:
            await write_bytes(accessor, path, data)
            return
        await write_bytes(accessor, path, existing + data)

    return append


def _make_path_write(fn: OpFn) -> OpFn:

    async def mutate(accessor: Accessor, path: PathSpec, **kwargs) -> None:
        await fn(accessor, path)

    return mutate


def _make_mkdir_parents(fn: OpFn, force_parents: bool = True) -> OpFn:

    async def mkdir(accessor: Accessor, path: PathSpec, **kwargs) -> None:
        await fn(accessor,
                 path,
                 parents=force_parents or kwargs.get("parents") is True)

    return mkdir


def _make_rename(fn: OpFn) -> OpFn:

    async def rename(accessor: Accessor, src: PathSpec, dst: PathSpec,
                     **kwargs) -> None:
        await fn(accessor, src, dst)

    return rename


def _make_truncate(fn: OpFn) -> OpFn:

    async def truncate(accessor: Accessor, path: PathSpec, length: int,
                       **kwargs) -> None:
        await fn(accessor, path, length)

    return truncate


def _make_emulated_truncate(read_bytes: OpFn, write_bytes: OpFn) -> OpFn:

    async def truncate(accessor: Accessor, path: PathSpec, length: int,
                       **kwargs) -> None:
        try:
            data = await read_bytes(accessor, path, index=NULL_INDEX)
        except FileNotFoundError:
            data = b""
        await write_bytes(accessor, path, data[:length].ljust(length, b"\0"))

    return truncate


def _make_set_attrs(fn: OpFn) -> OpFn:

    async def set_attrs(
        accessor: Accessor,
        path: PathSpec,
        *,
        mode: int | None = None,
        uid: int | str | None = None,
        gid: int | str | None = None,
        atime: str | None = None,
        mtime: str | None = None,
        index: IndexCacheStore | None = None,
        **kwargs,
    ) -> dict[str, int | str]:
        return await fn(accessor,
                        path,
                        mode=mode,
                        uid=uid,
                        gid=gid,
                        atime=atime,
                        mtime=mtime)

    return set_attrs


def _emit(ops: list[RegisteredOp], vfs_names: list[str], name: str, fn: OpFn,
          write: bool, filetype: str | None, overrides: set[str]) -> None:
    if name in overrides:
        return
    for res in vfs_names:
        ops.append(
            RegisteredOp(name=name,
                         vfs=res,
                         filetype=filetype,
                         fn=fn,
                         write=write))


def make_generic_ops(
    vfs: str | list[str],
    table: OpsTable,
    *,
    emulate_truncate: bool = False,
    mkdir_parents: bool = False,
    overrides: set[str] | None = None,
) -> list[RegisteredOp]:
    """Generate a backend's VFS/FUSE op set from its ``CommandIO`` table.

    The per-backend ``ops/<b>/`` wrapper modules were hand-written
    forwards of exactly five shapes; this factory emits the same
    wrappers from the table that already feeds
    ``make_generic_commands``, so a backend declares its core surface
    once. Ops whose table field is None are omitted, mirroring how the
    command factory skips write commands on read-only backends. A writable
    table without native append uses read-modify-write, which is async but
    not atomic against concurrent writers, like emulated truncate.

    ``index`` is forwarded into read/readdir/stat for every backend, so
    there is deliberately no ``forward_index`` knob here, in either
    language. Every write path evicts the parent listing through
    :meth:`Dispatcher.invalidate_after_write`, and both VFS/FUSE
    surfaces (:class:`Ops` in both languages) delegate to
    that same dispatcher, so the door that populates a listing is the
    door that evicts it.

    Every op emitted here is filetype-agnostic. To serve one extension
    differently, register a filetype-scoped op on the mount; the mount
    resolves ``(name, filetype)`` before ``(name, VFS)``.

    Args:
        vfs (str | list[str]): VFS name(s) the ops register
            under; a list fans out one ``RegisteredOp`` per name (the
            HF family registers one surface for four VFS).
        table (OpsTable): the backend's IO table (its ``CommandIO``).
        emulate_truncate (bool): synthesize ``truncate`` from
            ``read_bytes`` + ``write`` for a backend with no native
            partial write (dropbox is the only one; the rest grew a
            real ``truncate`` in their table, which wins outright).
        mkdir_parents (bool): forward ``parents=True`` to the core
            mkdir (disk).
        overrides (set[str] | None): op names to skip because the
            backend registers its own irregular wrapper.
    """
    vfs_names = vfs if isinstance(vfs, list) else [vfs]
    skip = overrides or set()
    ops: list[RegisteredOp] = []

    _emit(ops, vfs_names, "read", _make_ranged_read(table), False, None, skip)
    _emit(ops, vfs_names, "readdir", _make_read(table.readdir), False, None,
          skip)
    _emit(ops, vfs_names, "stat", _make_read(table.stat), False, None, skip)

    if table.write is not None:
        _emit(ops, vfs_names, "write", _make_data_write(table.write), True,
              None, skip)
    if table.append is not None:
        _emit(ops, vfs_names, "append", _make_data_write(table.append), True,
              None, skip)
    elif table.write is not None:
        _emit(ops, vfs_names, "append",
              _make_emulated_append(table.read_bytes, table.write), True, None,
              skip)
    if table.create is not None:
        _emit(ops, vfs_names, "create", _make_path_write(table.create), True,
              None, skip)
    if table.mkdir is not None:
        mkdir_fn = _make_mkdir_parents(table.mkdir, mkdir_parents)
        _emit(ops, vfs_names, "mkdir", mkdir_fn, True, None, skip)
    if table.unlink is not None:
        _emit(ops, vfs_names, "unlink", _make_path_write(table.unlink), True,
              None, skip)
    if table.rmdir is not None:
        _emit(ops, vfs_names, "rmdir", _make_path_write(table.rmdir), True,
              None, skip)
    if table.rename is not None:
        _emit(ops, vfs_names, "rename", _make_rename(table.rename), True, None,
              skip)

    if table.truncate is not None:
        _emit(ops, vfs_names, "truncate", _make_truncate(table.truncate), True,
              None, skip)
    elif emulate_truncate:
        if table.write is None:
            raise ValueError(
                "emulate_truncate requires a write op on the table")
        _emit(ops, vfs_names, "truncate",
              _make_emulated_truncate(table.read_bytes, table.write), True,
              None, skip)

    if table.set_attrs is not None:
        _emit(ops, vfs_names, "setattr", _make_set_attrs(table.set_attrs),
              True, None, skip)

    return ops
